VPC Service Controls demo pull request

mdgraboski opened the pull request:
This is my demo for creating a VPC Service Control environment and testing it out.  Please follow the steps in the README.md, and let me know what you think.

morgante commented on examples/vpc_service_control/README.md:
```diff
@@ -0,0 +1,129 @@
+# VPC Service Control Demo
```
Let's change this repo to `onprem_demo`.

mdgraboski replied:
Done.

Tfmenard replied:
I don't see the change. I don't think you push the commit for this change.

mdgraboski replied:
Done for real. :)

morgante commented on examples/vpc_service_control/README.md:
```diff
@@ -0,0 +1,129 @@
+3. Reserve a public IP for your "on-prem" VPN router (to be clear, further use of "on-prem" in these instructions refers to the GCP project that we're pretending is your on-prem environment).
+    - Open `onprem_project/main.tf` in your favorite text editor.
+    - Search for `STEP 1A`.
+    - Don't actually delete the line; just add `*/` to the **end** of the line.
```
Requiring commenting/uncommenting is a rather confusing workflow. Can you explain why we need this?

aaron-lane replied:
I agree. There are a lot of manual steps to this guide that make it daunting to follow. Could we not use more Terraform configuration to obviate the need for file edits and orchestration? Variables and remote states come to mind.

Tfmenard replied:
Here's a suggestion to take out all the commenting and uncommenting steps and embed some Terraform best practices in this demo.

Instead of a per project folder. Create folders per resource category. Here's an example folder structure:

examples/onprem_demo/projects (for all projects resources and to configure the [GCS backend](https://www.terraform.io/docs/backends/types/gcs.html)
examples/onprem_demo/network (for all networks, VPNs, firewall rules, routers, etc.. in both projects)
examples/onprem_demo/compute (for all VMs created in both projects)
examples/onprem_demo/vpc_sc ( for all VPC SC resources created in the VPC SC project)

Here's a [lab](https://github.com/Tfmenard/terraform-vpn-lab) I made from which you can reuse code

mdgraboski replied:
We need the commenting/uncommenting because we need to first reserve public IP's for both the on-prem and VPC SC projects, for the VPN router; then we need to "tell" each environment what the other environment's VPN router's IP is.  So onprem project depends on the IP in VPC SC project, and VPC SC project depends on the IP in onprem project.

Tfmenard replied:
You wouldn't need to do this if the VPNs in both projects were created in 1 root config/folder and you used implicit dependencies like I did [here](https://github.com/Tfmenard/terraform-vpn-lab/blob/6e613a31a6415fd5360192debe33b82c8e4c53e9/lab-vpn/network_vpn_gcp.tf#L49) in my lab.

aaron-lane commented on examples/vpc_service_control/README.md:
```diff
@@ -0,0 +1,129 @@
+7.  Continue bouncing back and forth between the two `main.tf` files, uncommenting code blocks in numeric order for `STEP 5A`, `6A`, `7A`, and running `terraform apply` after uncommenting each code block.
+    - `7A` is in `vpc_sc_project/main.tf`.  It will prompt you to enable the Cloud DNS API.  Enable it.
+
+TODO: Complete the remaining instructions with more detail.
```
It appears that this guide is not complete.

mdgraboski replied:
It's complete.  This was a suggestion to myself to add more detail, but I think it's sufficient.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/main.tf:
```diff
@@ -0,0 +1,318 @@
+/***************************************/
+/* STEP 8A: DELETE THIS COMMENT LINE TO ENABLE VPC SERVICE CONTROLS.  OR - USE THE SECURITY SECTION OF THE CLOUD CONSOLE TO ENABLE THEM  
+
+resource "google_access_context_manager_service_perimeter" "service_perimeter" {
```
This should be using the [regular_service_perimeter](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/tree/master/modules/regular_service_perimeter) submodule to be an example of the module.

mdgraboski replied:
Perhaps we can do this in release 2.

Tfmenard replied:
Unfortunately no, this a hard requirement for accepting this Pull request.
This repo is a VPC SC Terraform module. Examples should be at least show how to use the module.

mdgraboski replied:
Done.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,35 @@
+// you can get your access policy name (which is actually a number) by running:
+// `gcloud organizations list` <- to get your org number
+// `gcloud access-context-manager policies list --organization=<your org number>`   <- to get the access policy name
+variable "access_policy_name" {}
```
This assumes the policy is already created.
Consider using the [org policy module](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/blob/master) the same way the other examples use it. See [simple_example](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/blob/master/examples/simple_example/main.tf)

mdgraboski replied:
Done.

Tfmenard replied:
I think what you did is remove the comments but you didn't edit main.tf to use the org policy module in main.tf
Also, remember to add descriptions to all variables.

Tfmenard replied:
This change is still pending.
I think you may have used VPC Service Controls with the GUI once with another project and the default access policy was created as mentioned [here](https://cloud.google.com/access-context-manager/docs/create-access-policy)
You should either add a step to use a gcloud command to create the policy or use the org policy module in this repo.

mdgraboski replied:
Updated the README docs with the instructions on how to use an existing access policy, or to create one if it doesn't exist.

morgante commented:
@Tfmenard are you able to re-review this?

Tfmenard commented on examples/vpc_sc_perimeter_auto_add/README.md:
```diff
@@ -0,0 +1,25 @@
+- A valid billing account.
+- A service account with "Access Context Manager Admin" privileges.  
+- Terraform v0.11.13 with provider.google v2.6.0 (the only versions I've tested; run `terraform version` to see what you've got).
+
```
You can add the **Running the example** section here.
Here's a draft that's partly consistent with the other examples.

```suggestion
## Running the example

To provision this example, run the following from within this directory:
- `terraform init` to get the plugins
- `terraform plan` to see the infrastructure plan
- `terraform apply` to apply the infrastructure build
- `terraform destroy` to destroy the built infrastructure

### Cleaning up
- When you're done with this demo and have successfully deleted all the resources with `terraform destroy`, delete the service account keys stored locally and [delete the service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts#deleting_a_service_account) altogether from the Cloud Console.
```

mdgraboski replied:
Removed vpc_sc_perimeter_auto_add from this pull request, no longer relevant.

Tfmenard commented on examples/vpc_service_control/onprem_project/variables.tf:
```diff
@@ -0,0 +1,30 @@
+ */
+
+
+variable "project_id" {}
```
Pleas add descriptions to all variables and run `terraform fmt` to format all TF files

mdgraboski replied:
Done.

Tfmenard replied:
Maybe the commit wasn't pushed, but this variable and all other variables still don't have a description see [here](https://github.com/mdgraboski/terraform-google-vpc-service-controls/blob/44e3f3c84ae4741d275af0e5b458528dd1fb6ac9/examples/vpc_service_control/onprem_project/variables.tf#L18)

Tfmenard replied:
I just realized why you had written "Done" on all changess regarding variables descriptions.
This suggestion shows you where the description of the variable should be.
This is part of our standards and it's also used to autogenerate documentation.
Once you've removed the comments and added the description in the description move on to this change to autogenerate the documentation.
Once you've added 
```suggestion
variable "project_id" {
 description =  "The ID of the GCP project that is going to be created"
}
```

mdgraboski replied:
Done.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/main.tf:
```diff
@@ -0,0 +1,318 @@
+ * limitations under the License.
+ */
+
+provider "google" {
```
provider version should be pinned to 2.6.0

mdgraboski replied:
I don't know how!  :)

Tfmenard replied:
I showed how to do it [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r295053122)

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/main.tf:
```diff
@@ -0,0 +1,318 @@
+  region       = "${var.cloud_router_region}"
+}
+
+output "ip_addr_of_cloud_vpn_router" {
```
All outputs should be in a `outputs.tf` file not in the `main.tf` file.

mdgraboski replied:
I can't put it in outputs.tf because it's part of the commented-out block of code.  If I put it in outputs.tf, that's a new file where users will have to deal with commenting and uncommenting.  Makes it more difficult to use.

Tfmenard replied:
I understand your problem. Though, I think this is another example of the commenting/uncommenting workflow leading to bad Terraform practices. So again, I suggest changing the workflow and apply Terraform best practices as mentioned [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r295591024)

Tfmenard commented on examples/vpc_service_control/onprem_project/variables.tf:
```diff
@@ -0,0 +1,30 @@
+
+variable "billing_account_id" {}
+
+variable "cloud_router_region" {}
```
Since this variable is used pretty much everywhere not just for the router, it should simply be named region.
Also, a default for every variable that can have a default to reduce the setup time for users.
```suggestion
variable "region" {
description = "Region to deploy all resources in"
default = "us-central1"
}
```

mdgraboski replied:
Done.

Tfmenard replied:
The variable is still missing a default and a description.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/terraform.tfvars:
```diff
@@ -0,0 +1,8 @@
+billing_account_id = ""
+cloud_router_region = "us-west1"
+shared_secret_string_for_vpn_connection = ""
+ip_addr_of_onprem_vpn_router = ""
```
```suggestion
```

mdgraboski replied:
Will delete the tfvars file.

Tfmenard replied:
No need to delete the tfvars file. You can simply create a terraform.tfvars.example file as mentioned [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r295596474)

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,35 @@
+
+variable "cloud_router_region" {}
+
+variable "ip_addr_of_onprem_vpn_router" {}
```
With or without the commenting and uncommenting workflow, I don't think this variable is necessary nor is the output if the same name. 
```suggestion
```

mdgraboski replied:
We need this so the on-prem and VPC SC  VPN routers can find each other.

Tfmenard replied:
You're right I must have been looking in the wrong folder.

Tfmenard commented on examples/vpc_service_control/onprem_project/main.tf:
```diff
@@ -0,0 +1,285 @@
+  region       = "${var.cloud_router_region}"
+} 
+
+output "ip_addr_of_onprem_vpn_router" {
```
```suggestion
```

mdgraboski replied:
Why delete this line?  We need this info.  We take the IP address of the onprem VPN router, and we feed that value into the tfvars file for our VPC SC project.

Tfmenard replied:
I asked to delete as a follow up to my comment [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r295592261) outputs should be in outputs.tf file

Tfmenard commented on examples/vpc_service_control/onprem_project/variables.tf:
```diff
@@ -0,0 +1,30 @@
+
+variable "cloud_router_region" {}
+
+variable "ip_addr_of_cloud_vpn_router" {}
```
```suggestion
```

mdgraboski replied:
We need to keep this variable, so the VPC SC project knows how to talk to the on-prem VPN router.

Tfmenard replied:
Yeah this breaks your current setup. This suggestion was assuming you would have VPNs for both created in one root configs

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/main.tf:
```diff
@@ -0,0 +1,318 @@
+resource "google_compute_vpn_tunnel" "vpc_sc_vpn_tunnel" {
+  ike_version             = "2"
+  name                    = "vpc-sc-vpn-tunnel"
+  peer_ip                 = "${var.ip_addr_of_onprem_vpn_router}"
```
```suggestion
  peer_ip                 = "${google_compute_address.onprem_vpn_ip.address}"
```

mdgraboski replied:
This doesn't work, per my comment above.

Tfmenard replied:
Indeed. I should've added a comment saying this suggestion is assuming both VPNs are created in the same root config and won't work with your current setup.

Tfmenard commented on examples/vpc_service_control/onprem_project/main.tf:
```diff
@@ -0,0 +1,285 @@
+resource "google_compute_vpn_tunnel" "onprem_vpn_tunnel" {
+  ike_version             = "2"
+  name                    = "onprem-vpn-tunnel"
+  peer_ip                 = "${var.ip_addr_of_cloud_vpn_router}"
```
```suggestion
  peer_ip                 = "${google_compute_address.vpc_sc_vpn_ip.address}"
```

mdgraboski replied:
This suggested change causes an error, because `vpc_sc_vpn_ip` does not exist in this main.tf file; it exists in the other main.tf file for the VPC SC project.

`Error: resource 'google_compute_vpn_tunnel.onprem_vpn_tunnel' config: unknown resource 'google_compute_address.vpc_sc_vpn_ip' referenced in variable google_compute_address.vpc_sc_vpn_ip.address`

Tfmenard replied:
Indeed. I should've added a comment saying this suggestion is assuming both VPNs are created in the same root config and won't work with your current setup.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,35 @@
+ */
+
+
+variable "project_id" {}
```
Please descriptions to all variables

mdgraboski replied:
Done.

Tfmenard replied:
All variables still don't have descriptions. Maybe this commit wasn't pushed.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,35 @@
+
+variable "billing_account_id" {}
+
+variable "cloud_router_region" {}
```
```suggestion
variable "region" {}
```

mdgraboski replied:
Done.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,35 @@
+
+variable "ip_addr_of_onprem_vpn_router" {}
+
+variable "shared_secret_string_for_vpn_connection" {}
```
```suggestion
variable "vpn_shared_secret" {}
```

mdgraboski replied:
Done.

Tfmenard commented on examples/vpc_service_control/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,35 @@
+
+variable "shared_secret_string_for_vpn_connection" {}
+
+// you can get your access policy name (which is actually a number) by running:
```
Please remove these comments. This should be in the README

mdgraboski replied:
Done.

Tfmenard replied:
The comment removed, but I don't see the command referenced in the README.

Tfmenard commented on examples/vpc_service_control/onprem_project/main.tf:
```diff
@@ -0,0 +1,285 @@
+ * limitations under the License.
+ */
+
+provider "google" {
```
The provider version should be pinned to 2.6.0

mdgraboski replied:
How do I do this?  I don't see any documentation for specifying the provider version at https://www.terraform.io/docs/providers/google/provider_reference.html

Tfmenard replied:
I showed how to do it here
https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r295053122

Tfmenard commented on examples/vpc_service_control/onprem_project/variables.tf:
```diff
@@ -0,0 +1,30 @@
+
+variable "project_id" {}
+
+variable "service_account_file" {}
```
Change this variable name for consistency with the other examples
```suggestion
variable "credentials_path" {}
```

mdgraboski replied:
Done.

Tfmenard commented on examples/vpc_service_control/onprem_project/terraform.tfvars:
```diff
@@ -0,0 +1,7 @@
+project_id = ""
```
Please remove all terraform.tfvars.
If you want to provide an example terraform.tfvars create a terraform.tfvars.example file and add placeholder to in variables that need to be configured by the user. Then instruct in the README to rename the file to terraform.tfvars before running the example.

mdgraboski replied:
I think this limits the usefulness of the example.  Not every user has experience with Terraform, nor do they realize a terraform.tfvars file is necessary.  I'll remove the tfvars files if that's "the policy," but I don't think this helps people who are brand new to Terraform.

Tfmenard replied:
Yeah it;s kind of the "policy". it's actually best practices for CFT as the user should only upload terraform.tfvars files in their own repos that use the modules. I agree that a user may be unfamiliar with TF and not realize that it's necessary, but I think the alternative of using a terraform.tfvars.example is good compromise that shows the users that terraform.tfvars is necessary while not having to create a terraform.tfvars from scratch.

mdgraboski commented on examples/vpc_service_control/README.md:
```diff
@@ -0,0 +1,129 @@
+        - `terraform plan`
+            - It should show a plan to create one project, two routes, and one firewall rule.  
+        - `terraform apply` and type "yes" when it prompts you to create the resources.  Might take a few minutes to create the project (should be less than 5 minutes).
+        - You will get an error message saying the Compute Engine API isn't enabled.  There will be a URL in the error message.  Copy and paste the URL into a browser, and then enable the Compute Engine API (takes a few minutes; be patient, let the spinner spin).  Once it's enabled, run `terraform apply` again - it should succeed.
```
Let's save this for release 2.

Tfmenard replied:
Sure I don't passing over this. Just create an issue in the repo so we keep track of possible enhancement/bugs

Tfmenard replied:
Please create the issue and the link in this comment once done so I can resolve the convo.

mdgraboski replied:
Fixed this by updating the Terraform to enable the API's.

Tfmenard commented on examples/onprem_demo/main.tf:
```diff
@@ -0,0 +1,41 @@
+  region                       = "${var.region}"
+  vpn_shared_secret            = "${var.vpn_shared_secret}"
+  ip_addr_of_onprem_vpn_router = "${module.onprem_network.ip_addr_of_onprem_vpn_router}"
+  access_policy_name           = "1059294165525"
```
```suggestion
  access_policy_name           = "{var.access_policy_name}"
```

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/README.md:
```diff
@@ -0,0 +1,70 @@
+        - `billing_account_id`: the billing account that the project will charge to.
+        - `region`: region where the GCP resources will be created. 
+        - `vpn_shared_secret`: any random text string can go here.
+        - `access_policy_name`: Get this value by running the following two commands:
```
As per this [comment](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#discussion_r301367739) you should add a step to create the policy as a user may not have used VPC SC before. Or maybe a conditional step like "if `gcloud access-context-manager policies list --organization <Organization ID> ` returns nothing, create the policy"

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/README.md:
```diff
@@ -0,0 +1,70 @@
+- Play around with VPC Service Controls.
+- Remember to delete all resources when done.  Easiest way to do this is to run `terraform destroy`.
+
+When you're done with this demo, delete the service account keys and delete the service account altogether from the Cloud Console.
```
Adding these placeholders are necessary to create documentation for your example.
Once you've added them, install terraform-docs and run `make generate_docs` command from the root folder as described [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls#autogeneration-of-documentation-from-tf-files)


```suggestion
When you're done with this demo, delete the service account keys and delete the service account altogether from the Cloud Console.

[^]: (autogen_docs_start)
[^]: (autogen_docs_end)
```

mdgraboski replied:
Is there any way we can avoid this?  This seems like needless complexity.  Much easier for me to simply type some text in a README than to figure out what's going on in that autogen link.

Tfmenard replied:
No unfortunately this is using terraform-docs is our standard way of documenting all modules and examples in modules.
If you're having trouble making it work let me know. It shouldn't be harder than downloading terraform-docs, adding it to your path and running `make generate_docs`

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/README.md:
```diff
@@ -0,0 +1,70 @@
+    - Go to the Billing section of the Cloud Console.  Select the billing account you want to charge for the projects you create.  (If you're using your credit card, the default limit is 5 projects mapped to a billing account, in case project creation fails for you in the steps below).  On the right side of the screen, there's a "Permissions" section for your billing account.  Click "Add members, ...
+
+1. Create the Terraform variables file.
+    - In your favorite text editor, edit and save the following variables in `examples/onprem_demo/terraform.tfvars`:
```
```suggestion
    - In your favorite text editor, create and save a file named `terraform.tfvars` in the `examples/onprem_demo/` directory:
```

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/README.md:
```diff
@@ -0,0 +1,70 @@
+
+0. Set up the service account needed to run the Terraform code.
+    - In an existing project, create the service account.  Copy and paste its email address (the big, long, `gserviceaccount.com` address) to a scratch pad.  Download its .json secret key file into the `onprem_project/` and also copy it into the `vpc_sc_project` folders.  Be **extremely careful** with this file; delete it when you're done with this example.
+    - Go to the [IAM admin](https://console.cloud.google.com/iam-admin/iam) section of your *organization.*  (You should see your domain name at the top of the screen, not an individual project name).  Click Add.  In the "New Members" box, paste the email address of the service account you created.  Under the Roles dialog, search for "Project Creator."  Grant that role by clicking Save.  Your ser ...
```
```suggestion
    - Go to the [IAM admin](https://console.cloud.google.com/iam-admin/iam) section of your *organization.*  (You should see your domain name at the top of the screen, not an individual project name).  Click Add.  In the "New Members" box, paste the email address of the service account you created.  Under the Roles dialog, search for "Project Creator."  Grant that role by clicking Save.  Your service account can now create projects and act as a project owner for those projects.
    -  Add the additional Organization level permissions to the service account mentioned [here](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls#organization-level-permissions) to  make sure your service account can create VPC SC otherwise known as Access Context Manager resources.
```

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/README.md:
```diff
@@ -0,0 +1,70 @@
+0. Set up the service account needed to run the Terraform code.
+    - In an existing project, create the service account.  Copy and paste its email address (the big, long, `gserviceaccount.com` address) to a scratch pad.  Download its .json secret key file into the `onprem_project/` and also copy it into the `vpc_sc_project` folders.  Be **extremely careful** with this file; delete it when you're done with this example.
+    - Go to the [IAM admin](https://console.cloud.google.com/iam-admin/iam) section of your *organization.*  (You should see your domain name at the top of the screen, not an individual project name).  Click Add.  In the "New Members" box, paste the email address of the service account you created.  Under the Roles dialog, search for "Project Creator."  Grant that role by clicking Save.  Your ser ...
+    - Go to the Billing section of the Cloud Console.  Select the billing account you want to charge for the projects you create.  (If you're using your credit card, the default limit is 5 projects mapped to a billing account, in case project creation fails for you in the steps below).  On the right side of the screen, there's a "Permissions" section for your billing account.  Click "Add members, ...
```
```suggestion
    - Go to the Billing section of the Cloud Console.  Select the billing account you want to charge for the projects you create.  (If you're using your credit card, the default limit is 5 projects mapped to a billing account, in case project creation fails for you in the steps below).  On the right side of the screen, there's a "Permissions" section for your billing account.  Click "Add members," and paste the email address of the service account you created.  Grant the service account Billing Account User privileges, so that it can map the projects it creates to your billing account.  **Note the billing account ID** (the alphanumeric text separated by hyphens), as we'll use it below.
```

Tfmenard replied:
Change from Billing Admin to Billing Account User

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/README.md:
```diff
@@ -0,0 +1,70 @@
+            - `gcloud access-context-manager policies list --organization <Organization ID>` 
+2. From a command line terminal:
+    - `cd examples/onprem_demo`
+    - `terraform plan`
```
```suggestion
    - `terraform init`
    - `terraform plan`
```

mdgraboski replied:
Done.

Tfmenard commented:
When running the example I got the mentioned errors regarding Compute Engine API.
But then I got errors that wasn't listed as an expected error in the README.


```
Error: Error applying plan:

3 error(s) occurred:

* module.vpc_sc_network.module.regular_service_perimeter_1.google_access_context_manager_service_perimeter.regular_service_perimeter: 1 error(s) occurred:

* google_access_context_manager_service_perimeter.regular_service_perimeter: Error creating ServicePerimeter: googleapi: Error 403: The caller does not have permission
* module.onprem_network.google_compute_instance.windows_jumphost: 1 error(s) occurred:

* google_compute_instance.windows_jumphost: Error creating instance: googleapi: Error 400: Invalid value for field 'resource.networkInterfaces[0].networkIP': '10.138.0.3'. Requested internal IP is outside the subnetwork CIDR range., invalid
* module.onprem_network.google_compute_instance.forward_proxy_instance: 1 error(s) occurred:

* google_compute_instance.forward_proxy_instance: Error creating instance: googleapi: Error 400: Invalid value for field 'resource.networkInterfaces[0].networkIP': '10.138.0.2'. Requested internal IP is outside the subnetwork CIDR range., invalid

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
````

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/main.tf:
```diff
@@ -0,0 +1,44 @@
+ * limitations under the License.
+ */
+
+provider "google" {
```
Done.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/variables.tf:
```diff
@@ -0,0 +1,26 @@
+ * limitations under the License.
+ */
+
+variable "service_account_file" {}
```
Done.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/variables.tf:
```diff
@@ -0,0 +1,26 @@
+ * limitations under the License.
+ */
+
+variable "service_account_file" {}
```
Done.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/variables.tf:
```diff
@@ -0,0 +1,26 @@
+
+variable "service_account_file" {}
+
+// you can get your access policy name (which is actually a number) by running:
```
Done.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/terraform.tfvars:
```diff
@@ -0,0 +1,3 @@
+service_account_file = ".json"
```
Done, the tfvars files have been removed.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/terraform.tfvars:
```diff
@@ -0,0 +1,3 @@
+service_account_file = ".json"
```
Done.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/main.tf:
```diff
@@ -0,0 +1,44 @@
+  credentials = "${file("${var.service_account_file}")}"
+}
+
+resource "google_access_context_manager_service_perimeter" "service_perimeter" {
```
Done.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/README.md:
```diff
@@ -0,0 +1,25 @@
+4. Run `terraform apply` to update the service perimeter to include the newly-created GCP project.
+
+
+When you're done with this demo, delete the service account keys and delete the service account altogether from the Cloud Console.
```
This section of the README has been deleted.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/main.tf:
```diff
@@ -0,0 +1,44 @@
+
+    // Complete list of VPC Service Control services can be found at:
+    // https://cloud.google.com/vpc-service-controls/docs/supported-products#apis_and_service_perimeters
+    restricted_services = ["bigquery.googleapis.com",
```
The point was to include all possible services (as of today) and include them in the service perimeter.

Tfmenard replied:
Gotcha. Sure that's fine.

mdgraboski commented:
> When running the example I got the mentioned errors regarding Compute Engine API.
> But then I got errors that wasn't listed as an expected error in the README.
> 
> ```
> Error: Error applying plan:
> 
> 3 error(s) occurred:
> 
> * module.vpc_sc_network.module.regular_service_perimeter_1.google_access_context_manager_service_perimeter.regular_service_perimeter: 1 error(s) occurred:
> 
> * google_access_context_manager_service_perimeter.regular_service_perimeter: Error creating ServicePerimeter: googleapi: Error 403: The caller does not have permission
> * module.onprem_network.google_compute_instance.windows_jumphost: 1 error(s) occurred:
> 
> * google_compute_instance.windows_jumphost: Error creating instance: googleapi: Error 400: Invalid value for field 'resource.networkInterfaces[0].networkIP': '10.138.0.3'. Requested internal IP is outside the subnetwork CIDR range., invalid
> * module.onprem_network.google_compute_instance.forward_proxy_instance: 1 error(s) occurred:
> 
> * google_compute_instance.forward_proxy_instance: Error creating instance: googleapi: Error 400: Invalid value for field 'resource.networkInterfaces[0].networkIP': '10.138.0.2'. Requested internal IP is outside the subnetwork CIDR range., invalid
> 
> Terraform does not automatically rollback in the face of errors.
> Instead, your Terraform state file has been partially updated with
> any resources that successfully completed. Please address the error
> above and apply again to incrementally change your infrastructure.
> ```

Just updated examples/onprem_demo/onprem_project/main.tf to fix this error.

Tfmenard commented on examples/onprem_demo/variables.tf:
```diff
@@ -0,0 +1,39 @@
+ * limitations under the License.
+ */
+
+// The ID of the Onprem GCP project that is going to be created
```
You forgot to add description in this file

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/variables.tf:
```diff
@@ -0,0 +1,39 @@
+variable "billing_account_id" {}
+
+// GCP Region (like us-west1, us-central1, etc)
+variable "region" {}
```
Per our standards to make examples easier to run we add default variables where it's possible.
Please add to all other region variables as well in other variables.tf files.
```suggestion
variable "region" {
  description = "GCP Region"
  default = "us-central1"
}

```

mdgraboski replied:
Done.

Tfmenard commented on examples/vpc_sc_perimeter_auto_add/main.tf:
```diff
@@ -0,0 +1,44 @@
+  status {
+    resources = "${var.protected_projects_list}"
+
+    // Complete list of VPC Service Control services can be found at:
```
Actually never mind. Adding a link in variables description is not really clean.

Tfmenard commented:
Almost done with this PR :D.
One thing that wasn't address is this [comment](https://github.com/terraform-google-modules/terraform-google-vpc-service-controls/pull/4#pullrequestreview-251392241)
I think we should rethink this example and merge it in another PR.

mdgraboski commented:
@Tfmenard  I removed the VPC SC Perimeter auto-add example.  I also fixed the description field in the variables.tf files, and added a default region.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/README.md:
```diff
@@ -0,0 +1,25 @@
+## Prerequisites
+To complete this demo, you'll need:
+- A Google Cloud [organization](https://cloud.google.com/resource-manager/docs/quickstart-organizations) mapped to a domain that you own with at least one project already created.
+- A valid billing account.
```
Removed vpc_sc_perimeter_auto_add from this pull request, no longer relevant.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/README.md:
```diff
@@ -0,0 +1,25 @@
+
+This is a demo of adding a newly-created project to a [VPC Service Controls perimeter](https://cloud.google.com/vpc-service-controls/).  Currently, there's not a Google Cloud product-level feature that automatically adds a new project to a service perimeter; so we have to rely on automation to do this.
+
+## Prerequisites
```
Removed vpc_sc_perimeter_auto_add from this pull request, no longer relevant.

mdgraboski commented on examples/vpc_sc_perimeter_auto_add/README.md:
```diff
@@ -0,0 +1,25 @@
+To complete this demo, you'll need:
+- A Google Cloud [organization](https://cloud.google.com/resource-manager/docs/quickstart-organizations) mapped to a domain that you own with at least one project already created.
+- A valid billing account.
+- A service account with "Access Context Manager Admin" privileges.  
```
Removed vpc_sc_perimeter_auto_add from this pull request, no longer relevant.

Tfmenard commented on examples/onprem_demo/onprem_project/variables.tf:
```diff
@@ -0,0 +1,40 @@
+  default     = "us-west1"
+}
+
+variable "ip_addr_of_cloud_vpn_router" {
```
Try to keep variables name short.
```suggestion
variable "ip_addr_cloud_vpn" {
```

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/vpc_sc_project/variables.tf:
```diff
@@ -0,0 +1,44 @@
+  default     = "us-west1"
+}
+
+variable "ip_addr_of_onprem_vpn_router" {
```
Try to keep variables short free of articles
```suggestion
variable "ip_addr_onprem_vpn" {
```

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/onprem_project/outputs.tf:
```diff
@@ -0,0 +1,19 @@
+ * limitations under the License.
+ */
+
+output "ip_addr_of_onprem_vpn_router" {
```
Please add description to output for the sake o standards.

mdgraboski replied:
Done.

Tfmenard commented on examples/onprem_demo/vpc_sc_project/outputs.tf:
```diff
@@ -0,0 +1,19 @@
+ * limitations under the License.
+ */
+
+output "ip_addr_of_cloud_vpn_router" {
```
Please add description to output for the sake of standards

mdgraboski replied:
Done.

Tfmenard commented:
One last thing is missing that I forgot. The example should have an outputs.tf file to help the user navigate through the example while also giving them the confirmation that example completed sucessfully.
Here's a draft outputs.tf file of what I think the outputs.tf file could be. You can make it whatever as you know better, but there should be one. Outputting computed values is usually pretty useful.


outptus.tf file
```
/**
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

output "ip_addr_cloud_vpn_router" {
 description = "some description"
  value = "${module.onprem_network.ip_addr_cloud_vpn_router}"
}

output "ip_addr_onprem_vpn_router" {
 description = "some description"
  value = "${module.onprem_network.ip_addr_onprem_vpn_router}"
}

output "windows_jump_piblic_ip" {
 description = "some description"
value = "${module.onprem_network.windows_jump_public_ip}"
}
```

For adding the jumphost you would have to modify the onprem_project/outputs.tf file as well.
If you have the example already running try adding the outputs.tf file(s) and running terraform apply without running terraform destroy first. the outputs should show up in the logs directly.

mdgraboski commented:
I added the requested output file, which displays the external IP address of the Windows jumphost and the internal IP address of the VPC SC-protected Windows VM.
